### PR TITLE
Ideact listeners

### PR DESCRIPTION
**Pull request by stellatsv**
# Description of the PR
Listeners for the IDE activities:
- `IdeActionListener`: listens for actions triggered in the IDE (like run, debug, build, etc.)
- `ClassDeclarationListener`: listens to changes in the document and checks whether the class declaration is in the desired form.
- `FunctionDefinitionListener`: listens for changes in the document and checks if a specific method has the desired form.
- `OpenFileListener`: listens for files being opened in the IDE by the user (and also checks if they are already open).

The checks are based on the input given in the course configuration file.

An example that works with the `Football1` module would be:
```

"tutorials": {
  "24496": {
    "tasks": [
      {
        "instruction": "Now, fix the AddAwayGoal method's definition!",
        "component": "editor",
        "info": "Hint: there is a typo in the name",
        "action": "functionDefinition",
        "actionArguments": {
          "methodName": "addAwayGoal",
          "methodArguments": [],
          "methodBody": [
            "this.awayCount",
            "=",
            "this.awayCount",
            "+",
            "1"
          ],
          "filePath": "Football1/o1/football1/Match.scala",
          "typeParamClause": ""
        }
      },
      {
        "instruction": "Now, fix Match's declaration!",
        "component": "editor",
        "info": "Something is missing.",
        "action": "classDeclScala",
        "actionArguments": {
          "className": "Match",
          "classArguments": [
            "val home:A",
            "var away:int"
          ],
          "classHierarchy": "String",
          "traitHierarchy": [
            "String",
            "MatchTest"
          ],
          "typeParamClause": "[A]",
          "modifiers": [
            "sealed implicit",
            ""
          ],
          "annotations": [
            "@Nullable @Deprecated",
            ""
          ],
          "filePath": "Football1/o1/football1/Match.scala"
        }
      },
      {
        "instruction": "Now, fix the isHigherScoringThan method's definition!",
        "component": "editor",
        "info": "",
        "action": "functionDefinition",
        "actionArguments": {
          "methodName": "isHigherScoringThan",
          "methodArguments": [
            "anotherMatch:A"
          ],
          "methodBody": [
            "this.totalGoals",
            ">",
            "anotherMatch.totalGoals"
          ],
          "typeParamClause": "[A]",
          "filePath": "Football1/o1/football1/Match.scala"
        }
      },
      {
        "instruction": "Find and open the file Match.scala",
        "component": "projectTree",
        "info": "You can see your project's files here.",
        "action": "openEditor",
        "actionArguments": {
          "filePath": "Match.scala"
        }
      },
      {
        "instruction": "Now, fix the goalDifference method's definition!",
        "component": "editor",
        "info": "It should return Int",
        "action": "functionDefinition",
        "actionArguments": {
          "methodName": "goalDifference",
          "methodArguments": [],
          "methodBody": [
            "{",
            "this.homeGoals",
            "-",
            "this.awayGoals",
            "}"
          ],
          "filePath": "Football1/o1/football1/Match.scala",
          "typeParamClause": ""
        }
      },
      {
        "instruction": "Now, run MatchTest",
        "component": "editor",
        "info": "You can still see your project's files here.",
        "action": "run",
        "actionArguments": {
          "actionNames": [
            "Run...",
            "Run 'MatchTest'"
          ],
          "filePath": "Football1/o1/football1/MatchTest.scala"
        }
      },
      {
        "instruction": "Now, fix the isGoalless method's definition!",
        "component": "editor",
        "info": "It should return boolean",
        "action": "functionDefinition",
        "actionArguments": {
          "methodName": "isGoalless",
          "methodArguments": [],
          "methodBody": [
            "this.totalGoals",
            "==",
            "0"
          ],
          "filePath": "Football1/o1/football1/Match.scala",
          "typeParamClause": ""
        }
      }
    ]
  }
}



```

# Testing
<table>
  <tr>
    <th>unit</th>
    <th>integration</th>
    <th>e2e</th>
    <th>manual</th>
  </tr>
  <tr>
    <td>
      <ul>
        <li>- [x] new <b>unit</b> tests created</li>
        <li>- [x] all <b>unit</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>integration</b> tests created</li>
        <li>- [x] all <b>integration</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [ ] new <b>e2e</b> tests created</li>
        <li>- [x] all <b>e2e</b> tests pass</li>
      </ul>
    </td>
    <td>
      <ul>
        <li>- [x] <b>manual</b> testing went well</li>
      </ul>
    </td>
  </tr>
</table>  
  
# Have you updated the [TESTING.md](https://github.com/Aalto-LeTech/intellij-plugin/blob/master/TESTING.md) or other relevant documentation on _your_ branch?

- [ ] Yes.
- [ ] Not yet. I will do it next.
- [x] Not relevant.

### CODE REVIEW

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+
+  @Override
+  public void documentChanged(@NotNull DocumentEvent event) {
+    PsiDocumentManager.getInstance(project).commitDocument(event.getDocument());
+    PsiFile psiFile = PsiDocumentManager.getInstance(project).getPsiFile(event.getDocument());
+    checkPsiFile(psiFile);
+  }
+
+  private void checkPsiFile(PsiFile psiFile) {
+    if (psiFile != null) {
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:95]: move

Comment (jaakkonarhi) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:95]: please

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    }
+    boolean correct = true;
+    PsiParameter[] parameters = element.parameterList().getParameters();
+    for (PsiParameter param: parameters) {
+      if (!args.containsKey(param.getName())
+          || !args.get(param.getName()).equals(param.getType().getPresentableText())) {
+        correct = false;
+      }
+    }
+    return correct;
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:139]: Return early from the loop by returning false to reduce iterations and remove the unnecessary `correct` variable

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  private boolean checkExtendsBlock(ScExtendsBlockImpl element) {
+    List<String> hierarchies = new ArrayList<>();
+    PsiElement[] children = element.getChildren();
+    Optional<PsiElement> extendsElement = Arrays.stream(children).filter(
+        ScTemplateParentsImpl.class::isInstance).findFirst();
+    // Peculirarity: If 'extends' is written without specifying a class name
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:147]: Peculirarity: typo

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/FunctionDefinitionListener.java [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  }
+
+  private boolean checkParameters(PsiParameter[] parameters) {
+    Map<String, String> args = new HashMap<>();
+    for (String a: arguments) {
+      StringTokenizer tokenizer = new StringTokenizer(a, ":", false);
+      String key = tokenizer.nextToken();
+      String value = tokenizer.nextToken();
+      args.put(key, value);
+    }
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/FunctionDefinitionListener.java:129]: Extract this to a separate method

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IdeActionListener.java [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    }
+  }
+
+  @Override
+  public void beforeActionPerformed(@NotNull AnAction action, @NotNull DataContext dataContext,
+                                    @NotNull AnActionEvent event) {
+    //perhaps create subclasses?
+    boolean complete = true;
+    if (filePath != null && !filePath.isEmpty()) {
+      this.filePath = project.getBasePath() + filePath;
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IdeActionListener.java:59]: Consider using Path.of

File: src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java [hunk 6]
```diff
@@ -1,25 +1,55 @@
...
+  @Nullable Object get(@NotNull String key);
 
   /**
    * Gets an argument or throws a runtime exception if the argument is missing.
    * @param key Key
    * @return The (non-null) value of the argument.
    */
   default @NotNull String getOrThrow(@NotNull String key) {
-    String value = get(key);
+    if (! (get(key) instanceof String)) {
```
Comment (jaakkonarhi) [inline src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java:21]: Remove the controversial whitespace after exclamation mark please

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java (hunk 1 above)
Comment (OlliKiljunen) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:95]: I don't know what @jaakkonarhi means here, but you should annotate `psiFile` as `@Nullable`. And changing this to `if (psiFile == null) return;` makes the method easier to read.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+  }
+
+  private boolean checkScClass(ScClass element) {
+    return className.equals(element.getName());
+  }
+
+  private boolean checkScPrimaryConstructor(ScPrimaryConstructor element) {
+    Map<String, String> args = new HashMap<>();
+    for (String a: arguments) {
+      StringTokenizer tokenizer = new StringTokenizer(a, ":", false);
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:126]: I wasn't aware that Java has this useful class!

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+      String key = tokenizer.nextToken();
+      String value = tokenizer.nextToken();
+      args.put(key, value);
+    }
+    boolean correct = true;
+    PsiParameter[] parameters = element.parameterList().getParameters();
+    for (PsiParameter param: parameters) {
+      if (!args.containsKey(param.getName())
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:134]: I'd imagine the order of arguments being significant. It is not?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    Optional<PsiElement> extendsElement = Arrays.stream(children).filter(
+        ScTemplateParentsImpl.class::isInstance).findFirst();
+    // Peculirarity: If 'extends' is written without specifying a class name
+    // it is considered correct by Scala
+    // (the keyword extends is not visible when traversing the Psi tree)
+    if (extendsElement.isPresent()) {
+      if (hierarchy.length != 0) {
+        children = extendsElement.get().getChildren();
+        Arrays.stream(children).forEach(child -> hierarchies.add(child.getText()));
+        return hierarchies.equals(Arrays.asList(hierarchy));
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:154]: Here, apparently, the order matters. Is that wanted? Does it matter in Scala?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/FunctionDefinitionListener.java [hunk 10]
```diff
@@ -0,0 +1,164 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.plugins.scala.lang.psi.api.ScalaPsiElement;
+import org.jetbrains.plugins.scala.lang.psi.api.ScalaRecursiveElementVisitor;
+import org.jetbrains.plugins.scala.lang.psi.api.expr.ScAssignment;
+import org.jetbrains.plugins.scala.lang.psi.api.statements.ScFunctionDefinition;
+import org.jetbrains.plugins.scala.lang.psi.impl.expr.ScReferenceExpressionImpl;
+import org.jetbrains.plugins.scala.lang.psi.impl.statements.params.ScParametersImpl;
+
+
+public class FunctionDefinitionListener implements ActivitiesListener,
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/FunctionDefinitionListener.java:41]: Most of the logic in this class is the same as in `ClassDeclarationListener`. Reuse!

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IdeActionListener.java (hunk 5 above)
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IdeActionListener.java:59]: Don't reuse the class variable (which should be final). Introduce a local variable.

**Review by OlliKiljunen — COMMENTED** [review]
Nice work! I'd like to see a few changes, but even more, I think we should get clarity of which parts of our logic here we want to be independent of implementation details of IntelliJ and its PSI system.

Now, our listener classes has some kind of internal representations of program structures (class declarations, function declarations) they are expecting to see in the user code. When traversing the PSI, they then compare this internal structure to the PSI elements they find.

Consider an alternative approach: when config JSON is read, a specific internal representation object is created (let's say `ScalaClassDeclaration`) that contains the information what kind of class declaration we are expecting to see. Then, when listener finds a PSI class declaration with the same name, it constructs another `ScalaClassDeclaration` from the PSI data. Now the comparison of these two `ScalaClassDeclaration` objects can be done in a context that is independent of the IntelliJ PSI. Does this make sense to you?

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java (hunk 8 above)
Comment (jaakkonarhi) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java:134]: yes

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+import org.jetbrains.plugins.scala.lang.psi.api.toplevel.typedef.ScClass;
+import org.jetbrains.plugins.scala.lang.psi.impl.toplevel.templates.ScExtendsBlockImpl;
+import org.jetbrains.plugins.scala.lang.psi.impl.toplevel.templates.ScTemplateParentsImpl;
+
+public class ClassDeclarationListener implements ActivitiesListener, DocumentListener {
+
+  private final ListenerCallback callback;
+  private final Project project;
+  private final String className;
+  private String[] arguments;
```
[the remark this answers is not in the document]
Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ClassDeclarationListener.java]: It contains the arguments inside the constructor in the form of e.g. `name:String`, the format being the same as how the user would specify it in Scala code (with the aim to make the configuration file easier to read). Type parameters are specified separately by the variable `typeParameters` as they are represented separately in the PSI. Same goes for possible Modifiers and Annotations.

**Review by OlliKiljunen — CHANGES_REQUESTED** [review]
A challenging task well-tackled! :+1:

I left a few comments. I'm happy to re-review this later but please check those comments first.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+      }
+      return Arrays.equals(arguments, params.toArray());
+    }
+    return  false;
+  }
+
+  /**
+   * Checks the type parameters of the function.
+   * Certain characters are removed. In the case of type parameters, some text from the PSI
+   * comes in the form of [A, B] and so the brackets need to be removed.
```
[the remark this answers is not in the document]
Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java:79]: After looking into it again, turns out there is a simpler way to do this without so many string manipulations.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IntelliJActivityFactory.java [hunk 13]
```diff
@@ -21,7 +21,15 @@ public IntelliJActivityFactory(@NotNull Project project) {
...
+        return OpenFileListener.create(callback, project, arguments);
+      case "build":
+        return BuildActionListener.create(callback, project,arguments);
+      case "run":
+        return RunActionListener.create(callback, project, arguments);
+      case "classDeclScala":
+        return ClassDeclarationListener.create(callback, project, arguments);
+      case "functionDefinition":
+        return FunctionDefinitionListener.create(callback, project, arguments);
       default:
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/IntelliJActivityFactory.java:33]: Remark: I like this a lot now. We could even allow listeners to be added dynamically (by replacing switch-case with a search in Map<String, ListenerFactory> or such).

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ScalaCodeListener.java [hunk 14]
```diff
@@ -0,0 +1,78 @@
...
+import com.intellij.util.PathUtil;
+import fi.aalto.cs.apluscourses.model.task.ActivitiesListener;
+import fi.aalto.cs.apluscourses.model.task.ListenerCallback;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.concurrent.atomic.AtomicBoolean;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+public abstract class ScalaCodeListener implements DocumentListener, ActivitiesListener {
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/ScalaCodeListener.java:24]: This class seems not to be Scala specific so consider changing the name.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java [hunk 15]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.jetbrains.plugins.scala.lang.psi.api.base.ScAnnotations;
+import org.jetbrains.plugins.scala.lang.psi.api.base.ScPrimaryConstructor;
+import org.jetbrains.plugins.scala.lang.psi.api.base.types.ScSimpleTypeElement;
+import org.jetbrains.plugins.scala.lang.psi.api.statements.params.ScTypeParamClause;
+import org.jetbrains.plugins.scala.lang.psi.impl.base.ScModifierListImpl;
+import org.jetbrains.plugins.scala.lang.psi.impl.toplevel.templates.ScTemplateParentsImpl;
+
+public class ScalaClassDeclaration {
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java:23]: If we can create PSI elements from string literals (somehow like `PsiElement elem = PsiElement.fromString("class Foo() exteds Bar {}")`), this class could be easily and thoroughly unit-tested.

File: src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java [hunk 16]
```diff
@@ -1,25 +1,55 @@
...
 public interface Arguments {
-  @Nullable String get(@NotNull String key);
+  @Nullable Object get(@NotNull String key);
 
   /**
    * Gets an argument or throws a runtime exception if the argument is missing.
    * @param key Key
    * @return The (non-null) value of the argument.
    */
   default @NotNull String getOrThrow(@NotNull String key) {
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java:20]: I suggest renaming methods in this class:
- `getOrThrow` -> `getString`
- `getArrayOrThrow` -> `getArray`
- `get` -> `opt`

File: src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java [hunk 17]
```diff
@@ -1,25 +1,55 @@
...
+  default @NotNull String[] getArrayOrThrow(@NotNull String key) {
+    if (! (get(key) instanceof JSONArray)) {
+      throw new IllegalArgumentException("Argument is not a JSONArray: " + key);
+    }
+    JSONArray jsonArray = (JSONArray) get(key);
+    if (jsonArray == null) {
+      throw new IllegalArgumentException("Argument missing: " + key);
+    }
+
+    List<String> array = new ArrayList<>();
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java:45]: List named array (?)

File: src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java [hunk 18]
```diff
@@ -1,25 +1,55 @@
...
+    int length = jsonArray.length();
+    for (int i = 0; i < length; i++) {
+      array.add(jsonArray.getString(i));
+    }
+    return array.toArray(new String[]{});
+  }
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/model/task/Arguments.java:51]: You can use:
```
return JsonUtil.parseArray(jsonArray, JSONArray::getString, Function.identity(), String[]::new);
```
If that looks too ugly to you, you can overload `parseArray` that doesn't take `parser` parameter but passes `Function.identity()` in place of it. (Perhaps renaming that method would make sense in that case.)

File: src/main/java/fi/aalto/cs/apluscourses/utils/StringUtil.java [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+
+  private StringUtil() {}
+
+  /**
+   * Separates the given text into a String[] by tokenizing with the specified delimeter.
+   * @param phrase The String to be tokenized.
+   * @param delim The delimeter used in the tokenizing.
+   * @return The String[] containing the tokens.
+   */
+  public static String[] getArrayOfTokens(String phrase, char delim) {
```
Comment (OlliKiljunen) [inline src/main/java/fi/aalto/cs/apluscourses/utils/StringUtil.java:17]: The difference between this and `String::split` is still a bit unclear to me but ok. :D

File: src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/BuildActionListener.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+   */
+  public BuildActionListener(ListenerCallback callback, Project project,
+                             String[] action) {
+    super(callback, project, action);
+  }
+
+  public static BuildActionListener create(ListenerCallback callback, Project project,
+                                         Arguments arguments) {
+    return new BuildActionListener(callback, project,
+                arguments.getArrayOrThrow("actionNames"));
```
[the remark this answers is not in the document]
Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/model/task/BuildActionListener.java:24]: Checkstyle does not complain.

File: src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java (hunk 15 above)
Comment (stellatsv) [reply src/main/java/fi/aalto/cs/apluscourses/intellij/psi/ScalaClassDeclaration.java:23]: There is [this](https://github.com/JetBrains/intellij-community/blob/master/java/java-psi-impl/src/com/intellij/psi/impl/PsiElementFactoryImpl.java) promising class.

**Review by jaakkonarhi — APPROVED** [review]
If you press shift+shift, type Run and select the Run... with a green triangle, it completes the run task without running anything

**Pull request merged**